### DISCUSSION TITLE
Windows BarrierServer Setup

### DISCUSSION THREAD

**Discussion by skydiverscott** [debauchee/barrier#1616] [Q&A]
I am installing barrier in server mode 2.4.0 on Windows 10.  The gui install does not create the SSL cert for the server component.

After install using gui install, there is NO barrier/config file created and neither are the ssl cert files (.pem or fingerprint)

I have tried running this:
`MKDIR "%LocalAppData%\Barrier\SSL\Fingerprints" >NUL 2>&1
SET OPENSSL_CONF=C:\Program Files\Barrier\barrier.conf
SET RANDFILE=%LocalAppData%\Barrier\SSL\.rnd
"C:\Program Files\Barrier\openssl.exe" req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:2048 -keyout "%LocalAppData%\Barrier\SSL\Barrier.pem" -out "%LocalAppData%\Barrier\SSL\Barrier.pem"
IF EXIST "%RANDFILE%" DEL "%RANDFILE%"
FOR /F "tokens=2 delims=^=" %a in ('""C:\Program Files\Barrier\openssl.exe" x509 -fingerprint -sha1 -noout -in "%LocalAppData%\Barrier\SSL\Barrier.pem""') DO ECHO %a > "%LocalAppData%\Barrier\SSL\Fingerprints\Local.txt"`

But this doesn't work due to Windows not allowing writing to the C:\Program Files\Barrier directory

Can someone provide guidance on how to properly setup SSL cert on the server?
1. How can I create a basic barrier.conf file?
2. How can I manually create the barrier.pem file?

**Comment by llinfeng** [comment]
> 2. How can I manually create the barrier.pem file?

Recently, @ludificorpayne mentioned in [this issue](https://github.com/debauchee/barrier/issues/1625) that one way to get our hands on the `barrier.pem` file is to first install v2.3.4, and then install 2.4.0.

Actually, as [many](https://github.com/debauchee/barrier/issues/1377#issuecomment-994231023) [others](https://github.com/debauchee/barrier/issues/231) suggested, the following one-liner should suffice:
```
openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem
```

Though, per my case, I found the Git-version of `openssl` ended up producing a `barrier.pem` file that the Barrier client shall ignore. The solution is simple - use a WSL distro to execute `openssl` command gets the job done.

----

> 1. How can I create a basic barrier.conf file?

I don't have a good method to create the file. It took hours of tinkering to settle with the following configuration. It has been working well for years. (Though, to start from scratch, I found [the Synergy document](https://symless.com/help-articles/creating-text-config-files) very helpful.)
<details>
  <summary>Click to expand the configuration file, with A1-A5 belong to AW (server).</summary>
  
Here, Surface 7 and T480 are two clients. One runs Win 10, and one runs Win 11.

```
section: screens
	AW:
		halfDuplexCapsLock = false
		halfDuplexNumLock = false
		halfDuplexScrollLock = false
		xtestIsXineramaUnaware = false
		preserveFocus = false
		switchCorners = none 
		switchCornerSize = 0
	T480:
		halfDuplexCapsLock = false
		halfDuplexNumLock = false
		halfDuplexScrollLock = false
		xtestIsXineramaUnaware = false
		preserveFocus = false
		switchCorners = none 
		switchCornerSize = 0
	Surface7:
		halfDuplexCapsLock = false
		halfDuplexNumLock = false
		halfDuplexScrollLock = false
		xtestIsXineramaUnaware = false
		preserveFocus = false
		switchCorners = none 
		switchCornerSize = 0
end

section: aliases
end

#######################################
# A1-A4: 1200x1920                    #
#                                     #
#                        +---------+  #
#+-----+-----+-----+-----|         |  #
#|     |     |     |     |  A5     |  #
#+ A1  + A2  + A3  + A3  +---------+  #
#|     |     |     |     |            #
#+-----+-----+-----+-----+-------+    #
#        | Surface7 |     | T480  |    #
#        +---------+     +-------+    #
#######################################
section: links
	AW:
		down(40,60) = Surface7
		down(70,100) = T480
	T480:
		up = AW(80,100)
        left = Surface7
	Surface7:
		up(10,100) = AW(40,60)
		right(50,100) = T480
end

section: options
	relativeMouseMoves = false
	win32KeepForeground = false
	clipboardSharing = true
	switchCorners = none 
	switchCornerSize = 0
	keystroke(Shift+Alt+Control+&) = switchToScreen(Surface7)
    keystroke(Shift+Control+Alt+*) = switchToScreen(T480)
    keystroke(Shift+Alt+Control+() = switchToScreen(AW)
end
```
</details>
